distsql: pre-reserve memory needed to mark rows in HashJoiner build phase

asubiotto opened the pull request:
A situation was uncovered by #18600, where the HashJoiner would run out
of memory in the probe phase. This was because we had made an assumption
that we wouldn't hit a memory limit if the buffer phase filled up at
most 2/3 of the limit with both streams , since the marking
infrastructure would take up only a fraction of 1/3 (the chosen stream).

This assumption failed to take into account other limits shared with
other processors. This change pre-reserves the memory needed for the
probe phase in the build phase so that we can keep a single point in the
code where we fall back to disk while not relying on any limit
assumptions.

Commit:
distsql: pre-reserve memory needed to mark rows in HashJoiner build phase

A situation was uncovered by #18600, where the HashJoiner would run out
of memory in the probe phase. This was because we had made an assumption
that we wouldn't hit a memory limit if the buffer phase filled up at
most 2/3 of the limit with both streams , since the marking
infrastructure would take up only a fraction of 1/3 (the chosen stream).

This assumption failed to take into account other limits shared with
other processors. This change pre-reserves the memory needed for the
probe phase in the build phase so that we can keep a single point in the
code where we fall back to disk while not relying on any limit
assumptions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18838)
<!-- Reviewable:end -->

arjunravinarayan commented:
I'm not super happy with this PR, because it seems like a little more refactoring than absolutely necessary, but you should take my opinion with a grain of salt, because I'm working with a healthy amount of :dog: and am not sure I can articulate clearly whether you can make this PR more minimal. In either case, great job on fixing this bug. I'd like @andreimatei to weigh in before we merge this.

<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8CVfWPfEP387xwNU2:-Kv8CVfWPfEP387xwNU3:b1eq9t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e203bd9ac79ebfef32978a155f408a5e78cf1380/pkg/sql/distsqlrun/hashjoiner.go#L198)):*
> ```Go
> 	defer h.rows[rightSide].Close(ctx)
> 
> 	var storedRows hashRowContainer
> ```

Why rope in so much code-move refactoring into this commit (moving some amount of logic into `buildPhase`?) It's not entirely clear to me why it's necessary.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 305 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8D5768eCuK1esoJOB:-Kv8D5768eCuK1esoJOC:b-5luff) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L305)):*
> ```Go
> //    in-memory hashRowContainer representation.
> func (h *hashJoiner) buildPhase(
> 	ctx context.Context, useTempStorage bool, buildMemory bool,
> ```

s/`buildMemory`/`buildInMemory`/ makes this clearer, I was confused momentarily.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-:-Kv8QSMuE_uNMrSWOXAb:b-97y5z4)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Thanks for the review. I definitely thought about the amount of refactoring while writing this change but I think that it's small enough to be OK. The issue stems from the fact that I blurred the line between the buffer and build phases in a previous change: we build from disk if we encounter a memory error in the buffer phase and then in the level above, build from memory if we did not encounter a memory error.

The solution to this bug requires us to have a memHashRowContainer built so that we can pre-reserve the memory and fall back to disk if anything goes wrong. Thus the memHashRowContainer must be built before the diskHashRowContainer. This means either moving the in memory build phase to precede the disk build phase in the buffer phase or extract the disk build phase into a function that is called after building in-memory and where it currently is. I decided to clean up my previous mistake and instead move the disk build phase out of the buffer phase and put the build phase (of both memory and disk) behind a function. I think the code is better off this way and the change is small/localized enough that it will not introduce cherrypicking woes.

I have no issue with going with the other solution though so please let me know what you think.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I think the refactoring was quite necessary (and it's good we're cherry-picking it). Ideally I think it would have been better to do it in a different commit than the memory reservation change, but I shouldn't be the one to cast that stone :)

---

Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8ohUa2uHFrMvNHlFW:-Kv8ohUa2uHFrMvNHlFX:b-j2pyyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e203bd9ac79ebfef32978a155f408a5e78cf1380/pkg/sql/distsqlrun/hash_row_container.go#L122)):*
> ```Go
> // NOTE: Once a row is marked, adding more rows to the hashMemRowContainer
> // results in undefined behavior. It is not necessary to do otherwise for the
> // current usage of hashMemRowContainer and allows us to assume that a memory
> ```

Is the undefined behavior described by this note still necessary given that you're seemingly taking away the reason for it?

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 228 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8p71chKVsKyL-2i9d:-Kv8p71chKVsKyL-2i9e:b-v8zieb) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hash_row_container.go#L228)):*
> ```Go
> // memory account by the memory needed to mark all rows. It is a noop if
> // h.markMemoryReserved is true.
> func (h *hashMemRowContainer) reserveMarkMemory(ctx context.Context) error {
> ```

if the `!h.markMemoryReserved` condition is important, I'd rename this to `reserveMarkMemoryMaybe()`

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 229 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8p0DyUutWQZ94ozlw:-Kv8p0DyUutWQZ94ozlx:b-exxew4) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hash_row_container.go#L229)):*
> ```Go
> // h.markMemoryReserved is true.
> func (h *hashMemRowContainer) reserveMarkMemory(ctx context.Context) error {
> 	if !h.markMemoryReserved {
> ```

nit: invert the condition and return early so you can unident the bulk of this function.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 286 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv94ouomHSZyZ05w0Oz:-Kv94ouomHSZyZ05w0P-:b-fm7dpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hash_row_container.go#L286)):*
> ```Go
> 	}
> 	if i.marked == nil {
> 		if err := i.reserveMarkMemory(ctx); err != nil {
> ```

do we expect for this call to ever not be a no-op? If no, I'd find a way to assert that memory has already been reserved.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv91c3Iva51-n4Kooop:-Kv91c3Iva51-n4Koooq:bui60ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L210)):*
> ```Go
> 	// attempt to build an in-memory representation, falling back to disk
> 	// according to useTempStorage.
> 	storedRows, earlyExit, err := h.buildPhase(ctx, useTempStorage, err == nil /* buildMemory */)
> ```

I find everything around the use of this `err` very hard to read. The fact that `err != nil`at this point implies that it's a OOM error is quite obscure, and also the condition above for calling `DrainAndClose()` is a bit inscrutable. 
I suggest introducing a `oom bool` above.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv92amxRxktSl77pUEn:-Kv92amxRxktSl77pUEo:b-orfupl) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L223)):*
> ```Go
> 	defer storedRows.Close(ctx)
> 
> 	// Add the row that caused a memory error in the buffer phase if there is
> ```

Shouldn't this be moved above the  call to `buildPhase()`? Seems very randomly placed here.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 298 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv919yXIvZ0psdzF5h2:-Kv919yXIvZ0psdzF5h3:b-5h5062) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L298)):*
> ```Go
> 
> // buildPhase constructs our internal hash map of rows seen. This is done
> // entirely from one stream (chosen during initial buffering).
> ```

nit: hint to h.storedSide

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 303 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv9-p-BEj1v5r79KOml:-Kv9-p-BEj1v5r79KOmm:b-2o6zl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L303)):*
> ```Go
> //    storage if necessary.
> //  - buildMemory specifies whether the build phase should attempt to build an
> //    in-memory hashRowContainer representation.
> ```

as opposed to what? Consider giving a hint as to why this would ever be set to false - I guess it's when we already happen to know that there's no memory available. Then again, why not let the code always try to use memory and fallback?
Also, if we keep the arg, we should document and assert that useTempStorage is set when this is set.

Final nit: if we keep the arg, I'd suggest we name them `attemptMemoryBuild` and `fallbackToDiskBuild`.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 324 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv950U6dDRqwrUHoN-P:-Kv950U6dDRqwrUHoN-Q:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L324)):*
> ```Go
> 		// in-memory rows in this phase to not have to worry about hitting a
> 		// memory limit in the probe phase.
> 		err := storedMemRows.reserveMarkMemory(ctx)
> ```

:thumbsup: 

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 403 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8qDUmwXMpsvTuxSqR:-Kv8qDUmwXMpsvTuxSqS:b972kgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L403)):*
> bufferPhase attempts to read a portion

It doesn't "attempt to read", it reads :)

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8qHka72M3J-ghmx_8:-Kv8qHka72M3J-ghmx_9:b667xne) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L405)):*
> ```Go
> // bufferPhase attempts to read a portion of both streams into memory (up to
> // h.initialBufferSize) in the hope that one of them is small and should be used
> // as h.storedSide. The phase attempts to consume all the rows from the chosen
> ```

Is the last sentence finished?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-:-Kv95QxHPXy0JPlax4yl:b-aqnvpz)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 1 of 3 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8ohUa2uHFrMvNHlFW:-KvSZS3EZKi8LSfth2-3:bimcpvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e203bd9ac79ebfef32978a155f408a5e78cf1380/pkg/sql/distsqlrun/hash_row_container.go#L122)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Is the undefined behavior described by this note still necessary given that you're seemingly taking away the reason for it?
</blockquote></details>

The only reason there is undefined behavior is because it is not necessary to support adding more rows once we start marking. It is not necessary behavior, simply not worth time.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 228 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8p71chKVsKyL-2i9d:-KvS_sMpVM5VYYcJ3ZTf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hash_row_container.go#L228)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if the `!h.markMemoryReserved` condition is important, I'd rename this to `reserveMarkMemoryMaybe()`
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 229 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8p0DyUutWQZ94ozlw:-KvSalA3nm-yKwL_2Owm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hash_row_container.go#L229)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: invert the condition and return early so you can unident the bulk of this function.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 286 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv94ouomHSZyZ05w0Oz:-KvSeFhA1dnSJ_UFfn9P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hash_row_container.go#L286)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do we expect for this call to ever not be a no-op? If no, I'd find a way to assert that memory has already been reserved.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8CVfWPfEP387xwNU2:-KvSb9uuev1manDkgzv_:bgifv49) ([raw file](https://github.com/cockroachdb/cockroach/blob/e203bd9ac79ebfef32978a155f408a5e78cf1380/pkg/sql/distsqlrun/hashjoiner.go#L198)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Why rope in so much code-move refactoring into this commit (moving some amount of logic into `buildPhase`?) It's not entirely clear to me why it's necessary.
</blockquote></details>

Explained in main thread.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv91c3Iva51-n4Kooop:-KvSlZDhcyLZCZNLDHJh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L210)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I find everything around the use of this `err` very hard to read. The fact that `err != nil`at this point implies that it's a OOM error is quite obscure, and also the condition above for calling `DrainAndClose()` is a bit inscrutable. 
I suggest introducing a `oom bool` above.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv92amxRxktSl77pUEn:-KvSlm6A_tf0znEVfrFZ:but4ewm) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L223)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Shouldn't this be moved above the  call to `buildPhase()`? Seems very randomly placed here.
</blockquote></details>

This needs to be after `buildPhase` since we are adding the row to the built container. Rewrote the comment.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 298 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv919yXIvZ0psdzF5h2:-KvStG-qNs4C6NfkK1Ks:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L298)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: hint to h.storedSide
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 303 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv9-p-BEj1v5r79KOml:-KvSyltldhiZy0Qcue9o:bswbfmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L303)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

as opposed to what? Consider giving a hint as to why this would ever be set to false - I guess it's when we already happen to know that there's no memory available. Then again, why not let the code always try to use memory and fallback?
Also, if we keep the arg, we should document and assert that useTempStorage is set when this is set.

Final nit: if we keep the arg, I'd suggest we name them `attemptMemoryBuild` and `fallbackToDiskBuild`.
</blockquote></details>

Updated comment.

Since we already have knowledge of whether we should fall back to disk or not (disregarding marking), why not use that knowledge? I prefer to keep this as it is.

re assertions I'm not sure what you mean. That we assert that `useTempStorage` is true when `buildMemory` is true? It's a valid case to try to build in memory, try fall back to disk, and see that external storage is disabled.

Renamed `buildMemory` but I think `fallbackToDiskBuild` is more confusing than `useTempStorage`. This is what the boolean is called at the caller level as well (where we figure out whether we can use temp storage)

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 305 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8D5768eCuK1esoJOB:-KvT0q2I92qJj6dhsosU:b13gejq) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L305)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

s/`buildMemory`/`buildInMemory`/ makes this clearer, I was confused momentarily.
</blockquote></details>

Renamed to `attemptMemoryBuild`

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 403 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8qDUmwXMpsvTuxSqR:-KvT0xaqxE6WpH9X0XKB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L403)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> bufferPhase attempts to read a portion

It doesn't "attempt to read", it reads :)

</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838#-Kv8qHka72M3J-ghmx_8:-KvT1Z8UN8I7tg8GC3my:bjhrbs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/51a8153ff04df55282c50812673473e2300c5d64/pkg/sql/distsqlrun/hashjoiner.go#L405)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Is the last sentence finished?
</blockquote></details>

Yes, updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18838)*
<!-- Sent from Reviewable.io -->
